pkg/vmo/statefulset.go

```
@@ -75,3 +88,46 @@ func CreateStatefulSets(controller *Controller, vmo *vmcontrollerv1.VerrazzanoMo
 	glog.V(4).Infof("Successfully applied StatefulSets for vmo '%s'", vmo.Name)
 	return nil
 }
+
+// Update the each PVC owner reference to be the StatefulSet (STS).
+// PVCs are created automatically by Kubernetes when the STS is created
+// because the STS has a volumeClaimTemplate.  However, the PVCs are not deleted
+// when the STS is deleted. Set the PVC owner reference to be the STS
+// so that when the STS is deleted, the PVC will automatically get deleted.
+// Because PVC is dynamic, when it is deleted, the bound PV will also get deleted.
+// NOTE: This cannot be done automatically using the STS VolumeClaimTemplate.
+func updateOwnerForPVCs(controller *Controller, statefulSet *appsv1.StatefulSet, vmoName string, vmoNamespace string) error {
+
+	// Get the PVCs for this STS using the specID label. Each PVC metadata.label
+	// has the same specID label as the STS template.metadata.label.
+	// For example: " app: hello-world-binding-es-master"
+	idLabel := resources.GetSpecID(vmoName, config.ElasticsearchMaster.Name)
+	selector := labels.SelectorFromSet(idLabel)
+	existingPvcList, err := controller.pvcLister.PersistentVolumeClaims(vmoNamespace).List(selector)
+	if err != nil {
+		return err
+	}
+	for _, pvc := range existingPvcList {
+		if len(pvc.OwnerReferences) != 0 {
+			continue
+		}
+		pvc.OwnerReferences = []metav1.OwnerReference{{
+			APIVersion: "apps/v1",
+			Kind:       "StatefulSet",
+			Name:       statefulSet.Name,
+			UID:        statefulSet.UID,
+		}}
+		glog.V(4).Infof("Setting StatefuleSet owner reference for PVC %s", pvc.Name)
+		_, err := controller.kubeclientset.CoreV1().PersistentVolumeClaims(vmoNamespace).Update(context.TODO(), pvc, metav1.UpdateOptions{})
+		if err != nil {
+			glog.Errorf("Failed to update the owner reference in PVC %s, for the reason (%v)", pvc.Name, err)
+			return err
+		}
+	}
+	replicas := int(*statefulSet.Spec.Replicas)
+	numPVCs := len(existingPvcList)
+	if numPVCs != replicas {
```

How could this happen?

---

It could be because the pvclister uses a read cache and hasn't been updated with all PVCs that are created asynchronously.  I actually saw this while testing so I added this check.

---

Should the pvclister not be used in this case so it does not rely on the cache and a error is not reported

---

Actually that should not be an error, i can fix that. In practice it is rarely seen, i tested several times.  However I did see it once.

---

I think using the lister is ok since this is the normal pattern and is more efficient than calling the API.  Also, I am not positive the lister is the issue since PVCs are created async after the STS is created.